src/Microsoft.AspNetCore.Html.Abstractions/HtmlFormattableString.cs

```diff
@@ -0,0 +1,179 @@
+        {
+            if (format == null)
+            {
+                throw new ArgumentNullException(nameof(format));
+            }
+
+            if (args == null)
+            {
+                throw new ArgumentNullException(nameof(args));
+            }
+
+            _formatProvider = formatProvider ?? CultureInfo.CurrentCulture;
+            _format = format;
+            _args = args;
+        }
+
+        /// <inheritdoc />
+        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            if (encoder == null)
+            {
+                throw new ArgumentNullException(nameof(encoder));
+            }
+
+            var formatProvider = new EncodingFormatProvider(_formatProvider, encoder);
+            writer.Write(string.Format(formatProvider, _format, _args));
+        }
+
+        private string DebuggerToString()
+        {
+            using (var writer = new StringWriter())
+            {
+                WriteTo(writer, HtmlEncoder.Default);
+                return writer.ToString();
+            }
+        }
+
+        // This class implements Html encoding via an ICustomFormatter. Passing an instance of this
+        // class into a string.Format method or anything similar will evaluate arguments implementing
+        // IHtmlContent without HTML encoding them, and will give other arguments the standard
+        // composite format string treatment, and then HTML encode the result.
+        //
+        // Plenty of examples of ICustomFormatter and the interactions with string.Format here:
+        // https://msdn.microsoft.com/en-us/library/system.string.format(v=vs.110).aspx#Format6_Example
+        private class EncodingFormatProvider : IFormatProvider, ICustomFormatter
+        {
+            private readonly HtmlEncoder _encoder;
+            private readonly IFormatProvider _formatProvider;
+
+            public EncodingFormatProvider(IFormatProvider formatProvider, HtmlEncoder encoder)
+            {
+                Debug.Assert(formatProvider != null);
+                Debug.Assert(encoder != null);
+
+                _formatProvider = formatProvider;
+                _encoder = encoder;
+            }
+
+            public string Format(string format, object arg, IFormatProvider formatProvider)
+            {
+                // These are the cases we need to special case. We trust the HtmlEncodedString or IHtmlContent instance
+                // to do the right thing with encoding.
+                var htmlString = arg as HtmlEncodedString;
+                if (htmlString != null)
+                {
+                    return htmlString.ToString();
+                }
+
+                var htmlContent = arg as IHtmlContent;
+                if (htmlContent != null)
+                {
+                    using (var writer = new StringWriter())
```

Comment:
Suggest giving this `class` a lazy-init `StringWriter` property that this method clears after each use.